### PR TITLE
fix: environ might be missing

### PR DESCRIPTION
**Pull request by kaporzhu**
how it happened?
the environ is attached to WSGIRequestHandler object in run_wsgi method. But run_wsgi is only executed after the request is pasted in handle_one_request method. If request is pasted failed and try to send_error message. It will try to call address_string method to get the client address and raise the AttributeError(AttributeError: 'WSGIRequestHandler' object has no attribute 'environ')

### CODE REVIEW

**Comment by andrewtchin** [comment]
:+1: please merge

**Comment by bhgames** [comment]
yes having this issue too

File: werkzeug/serving.py [hunk 1]
```diff
@@ -263,7 +263,7 @@ def version_string(self):
         return BaseHTTPRequestHandler.version_string(self).strip()
 
     def address_string(self):
-        return self.environ['REMOTE_ADDR']
+        return self.environ['REMOTE_ADDR'] if self.environ else 'localhost'
```
Comment (untitaker) [inline werkzeug/serving.py]: You say you recieve an AttributeError, but `if self.environ` will fail the same way.

Comment (kaporzhu) [inline werkzeug/serving.py]: Oh, good point, my mistake. it should be 

    getattr(self, 'environ', None)

or we should init the **environ** in the `__init__`

**Comment by idanya** [comment]
Having the same issue. Please merge.

**Comment by untitaker** [comment]
I will not merge this PR yet because it contains a few unrelated changes and
one nitpick is not resolved.

Stop the mindless +1.

**Comment by ctoth** [comment]
These are hardly mindless +1's. these are real people who are really 
having real trouble with this real problem.
Please fix it, or if you won't accept that pr explain what you want 
precisely so someone can get it done *today*.
  as is the sort of mindless focus on some arbitrary metric is far more 
destructive than simply not fixing this issue, and then calling people 
mindless because it hits them.

**Comment by untitaker** [comment]
>if you won't accept that pr

I said **yet**. It doesn't help a bit to post :+1: s while the PR is in this state.

>explain what you want precisely

* I left a comment on the relevant line in the diff. The author has responded back, agreeing.

* Then some other commits for totally unrelated issues got pushed into the same branch.

* Please understand that Flask and Werkzeug get what feels like an endless amount of trivial one-line PRs, all supposedly fixing a problem the author has. They're often the wrong fix to a "bug" in some other library, or a workaround to incorrect usage of the APIs by the PR author. In other words, the PRs are utter crap. To be fair, this doesn't seem to be the case here. The explanation the author gave seems somewhat logical, but please still write a testcase for this so I can sanely verify this bug.

**Comment by davidalbertonogueira** [comment]
Can't you just commit a line change in werkzeug\serving.py and submit the unrelated changes in another commit?
werkzeug\serving.py : line 266
`    def address_string(self):
        return self.environ['REMOTE_ADDR'] if getattr(self, 'environ', None) else 'localhost'`

**Comment by idanya** [comment]
Lets continue talking about this one line and not merge it... Maybe the bug will get tired and stop raising exceptions.

**Comment by untitaker** [comment]
Let's ignore what I wrote above too.

**Comment by patjones80** [comment]
Good evening,

It seems that there is some back and forth going on here that I'll happily stay out of, except to say that I'm getting **AttributeError: 'WSGIRequestHandler' object has no attribute 'environ'** under Werkzeug 0.11.4 and Flask 0.10.1. The context is that I'm trying to force HTTPS using SSLify. I have not seen this issue prior to installing SSLify and trying this. 

On a whim I modified werkzeug/serving.py as indicated above by @untitaker, and by @nathancahill in his commit, but that only made the traceback even more indecipherable. I'm guessing that I need this modification as part of a completely new install of Werkzeug?

Comments appreciated, thanks so much.

**Comment by JWKennington** [comment]
mindless +1

Also having this error, would appreciate any insight into manual solutions or a roadmap for how this bug will be squashed. 

Thanks!

**Comment by untitaker** [comment]
Let's see what we have:

* One pull request that contains a fix for the apparently real, but insufficiently described issue, see #873 
* This pull request that contains a broken fix and a lot of unrelated changes.
* A lot of mindless :+1:s
* **An extremely popular bug without anybody trying to present a reproducible testcase or example application**.

**Comment by JWKennington** [comment]
petulance aside, I agree that the issue is abnormally absent of any MWE. Unfortunately, most of the instances in which I experience the issue involve foreign-born requests attempting csrf on my flask application. Try reproducing that without a lot of overhead!

In an attempt to be helpful (and hopefully lower your blood pressure a tad untitaker), here's what I do know:
* anaconda python 3.5 64 bit
* werkzeug 0.11.3
* flask 0.10.1
* vanilla ubuntu 14.04 LTS setup on digital ocean, no peripherals


I'll provide a traceback since the error is nearly entirely outside the scope of my application, and may thus be not entirely useless:

_Exception happened during processing of request from ('37.187.152.107', 51623)
Traceback (most recent call last):
  File "/home/dev/anaconda3/lib/python3.5/socketserver.py", line 318, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/home/dev/anaconda3/lib/python3.5/socketserver.py", line 344, in process_request
    self.finish_request(request, client_address)
  File "/home/dev/anaconda3/lib/python3.5/socketserver.py", line 357, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/home/dev/anaconda3/lib/python3.5/socketserver.py", line 684, in __init__
    self.handle()
  File "/home/dev/anaconda3/lib/python3.5/site-packages/werkzeug/serving.py", line 217, in handle
    rv = BaseHTTPRequestHandler.handle(self)
  File "/home/dev/anaconda3/lib/python3.5/http/server.py", line 415, in handle
    self.handle_one_request()
  File "/home/dev/anaconda3/lib/python3.5/site-packages/werkzeug/serving.py", line 251, in handle_one_request
    elif self.parse_request():
  File "/home/dev/anaconda3/lib/python3.5/http/server.py", line 327, in parse_request
    "Bad request syntax (%r)" % requestline)
  File "/home/dev/anaconda3/lib/python3.5/http/server.py", line 445, in send_error
    self.log_error("code %d, message %s", code, message)
  File "/home/dev/anaconda3/lib/python3.5/site-packages/werkzeug/serving.py", line 273, in log_error
    self.log('error', *args)
  File "/home/dev/anaconda3/lib/python3.5/site-packages/werkzeug/serving.py", line 279, in log
    _log(type, '%s - - [%s] %s\n' % (self.address_string(),
  File "/home/dev/anaconda3/lib/python3.5/site-packages/werkzeug/serving.py", line 267, in address_string
    return self.environ['REMOTE_ADDR']
AttributeError: 'WSGIRequestHandler' object has no attribute 'environ'_

**Comment by tlynn** [comment]
Please merge [nathancahill's fix](https://github.com/nathancahill/werkzeug/commit/5953e59ca2f7f13104964245db68f3062fd707e1). Whether or not it's correct (it appears correct to me), it definitely addresses a statically-detectable bug in the code, and here's the evidence that it's a real-world improvement:
```sh
$ cd /tmp
$ virtualenv wz838env
$ ./wz838env/bin/pip install Flask
   ...
Successfully installed Flask-0.10.1 Jinja2-2.8 MarkupSafe-0.23 Werkzeug-0.11.4 itsdangerous-0.24
$ cat >hello.py
from flask import Flask
app = Flask(__name__)

@app.route("/")
def hello():
    return "Hello World!"

if __name__ == "__main__":
    app.run()
$ ./wz838env/bin/python hello.py &
[1] 7297
* Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
$ curl https://127.0.0.1:5000/
----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 35302)
Traceback (most recent call last):
  ...
  File "/tmp/wz838env/local/lib/python2.7/site-packages/werkzeug/serving.py", line 266, in address_string
    return self.environ['REMOTE_ADDR']
AttributeError: 'WSGIRequestHandler' object has no attribute 'environ'
----------------------------------------
curl: (35) Unknown SSL protocol error in connection to 127.0.0.1:5000 
$ kill %1
[1]+  Terminated              ./wz838env/bin/python hello.py
$ cat fix838.patch
diff -ur a/wz838env/lib/python2.7/site-packages/werkzeug/serving.py b/wz838env/lib/python2.7/site-packages/werkzeug/serving.py
--- a/wz838env/lib/python2.7/site-packages/werkzeug/serving.py	2016-03-22 16:08:39.917903000 +0000
+++ b/wz838env/lib/python2.7/site-packages/werkzeug/serving.py	2016-03-22 16:16:54.605901875 +0000
@@ -263,7 +263,10 @@
         return BaseHTTPRequestHandler.version_string(self).strip()
 
     def address_string(self):
-        return self.environ['REMOTE_ADDR']
+        if getattr(self, 'environ', None):
+            return self.environ['REMOTE_ADDR']
+        else:
+            return ''
 
     def log_request(self, code='-', size='-'):
         self.log('info', '"%s" %s %s', self.requestline, code, size)
$ cat fix838.patch | (cd wz838env/ && patch -p2)
$ ./wz838env/bin/python hello.py &
$ curl https://127.0.0.1:5000/
 - - [22/Mar/2016 16:23:53] code 400, message Bad request version ('X\x93(kI')
 - - [22/Mar/2016 16:23:53] "
                                 �m�1"�29T\f)��	X�(kI" 400 -
curl: (35) error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol
```

**Comment by untitaker** [comment]
@tlynn Thank you, that's an example I can reproduce.

**Comment by untitaker** [comment]
Also, y'all are using the development server in production, which is not a good idea. See the Flask deployment docs in master for a detailed explanation.

**Comment by untitaker** [comment]
0.11.5 is out with this fix.

**Pull request closed without merging**